File: pc8top86.py
import os
import struct
import sys

def create_p86(input_files, output_file, p86drv_version=0x11):
    pc8_data_list = []
    all_size = 0x10  # Initialize all_size with 0x10

    # Read all input files and calculate the total size
    for input_file in input_files:
        with open(input_file, "rb") as pc8_file:
            pc8_data = pc8_file.read()
        pc8_data_list.append(pc8_data)
        all_size += len(pc8_data)  # Add each file's size to all_size

    all_size += 0x6 # I believe this is for the size of the header ?

    # Prepare the header
    header = bytearray()
    header.extend(b"PCM86 DATA\x0A\x00")  # 12 bytes
    header.append(p86drv_version)  # Version number 1 byte
    header.extend(all_size.to_bytes(3, 'little'))  # ALL_Size 3 bytes

    # Set START and SIZE for each input file
    start = 0x610
    for pc8_data in pc8_data_list:
        header.extend(start.to_bytes(3, 'little'))  # START (3 bytes)
        header.extend(len(pc8_data).to_bytes(3, 'little'))  # SIZE (3 bytes)
        start += len(pc8_data)

    # Fill the rest of the START/SIZE values with zeros
    header.extend(b'\x00\x00\x00' * (255 * 2 - len(pc8_data_list) * 2))  # START/SIZE (3 bytes each) 6 bytes * (255 - number of input files)

    # Calculate padding size
    padding_size = 0x610 - len(header)

    # Add padding (empty bytes) to header
    header.extend(b'\x00' * padding_size)

    with open(output_file, "wb") as p86_file:
        p86_file.write(header)
        for pc8_data in pc8_data_list:
            p86_file.write(pc8_data)

if __name__ == "__main__":
    if len(sys.argv) < 3:
        print("Usage: python3 create_p86.py <input_file1> <input_file2> ... <output_file>")
        sys.exit(1)

    input_files = sys.argv[1:-1]
    output_file = sys.argv[-1]
    create_p86(input_files, output_file)

File: pc8top86.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>

static uint8_t *pc8_data_list[256];
static uint32_t pc8_data_sizes[256];

int create_p86(char *input_files[], int input_files_count, char *output_file, uint8_t p86drv_version) {
	int i;
	uint32_t all_size = 0x10; // Initialize all_size with 0x10
	uint8_t header[0x610];
	FILE *pc8_file;
	uint32_t start;
	FILE *p86_file;
	
    // Read all input files and calculate the total size
    for (i = 0; i < input_files_count; i++) 
    {
		pc8_file = fopen(input_files[i], "rb");
        if (!pc8_file) {
            puts("Error opening input file\n");
            return 1;
        }

        fseek(pc8_file, 0, SEEK_END);
        pc8_data_sizes[i] = ftell(pc8_file);
        fseek(pc8_file, 0, SEEK_SET);

        pc8_data_list[i] = malloc(pc8_data_sizes[i]);
        fread(pc8_data_list[i], 1, pc8_data_sizes[i], pc8_file);
        fclose(pc8_file);

        all_size += pc8_data_sizes[i]; // Add each file's size to all_size
    }

    all_size += 0x6; // Add 6 bytes for the size of the header

    // Prepare the header
    memcpy(header, "PCM86 DATA\x0A\x00", 12);
    header[12] = p86drv_version;
    memcpy(header + 13, &all_size, 3);

    memset(header + 16, 0, 6); // Set the first 6 bytes of START/SIZE to zero (index 0 is reserved)

    // Set START and SIZE for each input file
    start = 0x610;
    for (i = 0; i < input_files_count; i++) {
        memcpy(header + 22 + i * 6, &start, 3);
        memcpy(header + 25 + i * 6, &pc8_data_sizes[i], 3);
        start += pc8_data_sizes[i];
    }

    // Fill the rest of the START/SIZE values with zeros
    memset(header + 22 + input_files_count * 6, 0, (255 * 2 - input_files_count * 2) * 3);

    // Write the output file
    p86_file = fopen(output_file, "wb");
    if (!p86_file) {
        perror("Error opening output file");
        exit(1);
    }

    fwrite(header, 1, sizeof(header), p86_file);
    for (i = 0; i < input_files_count; i++) {
        fwrite(pc8_data_list[i], 1, pc8_data_sizes[i], p86_file);
        free(pc8_data_list[i]);
    }

    fclose(p86_file);
    
    return 0;
}

int main(int argc, char *argv[]) 
{
	int input_files_count;
	char **input_files;
	char *output_file;
	
    if (argc < 3) {
        puts("Usage: pc8top86 input1.pc8 [input2.pc8 ...] output.p86\n");
        return 1;
    }

	input_files_count = argc - 2;
	input_files = argv + 1;
	output_file = argv[argc - 1];
	
	if (input_files_count > 255)
	{
		puts("Too many input files !\n Limit is 255\n");
		return 1;
	}

    return create_p86(input_files, input_files_count, output_file, 0x11);
}